src/circuit/gadget/sinsemilla/merkle/chip.rs

```diff
@@ -0,0 +1,520 @@
+    pub(super) sinsemilla_config: SinsemillaConfig,
+}
+
+#[derive(Clone, Debug)]
+pub struct MerkleChip {
+    config: MerkleConfig,
+}
+
+impl Chip<pallas::Base> for MerkleChip {
+    type Config = MerkleConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl MerkleChip {
+    pub fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        sinsemilla_config: SinsemillaConfig,
+    ) -> MerkleConfig {
+        let advices = sinsemilla_config.advices();
+        let cond_swap_config = CondSwapChip::configure(
+            meta,
+            advices[..5].try_into().unwrap(),
+            sinsemilla_config.perm.clone(),
+        );
+        let lookup_config = LookupRangeCheckConfig::configure(
+            meta,
+            advices[0],
+            sinsemilla_config.generator_table.table_idx,
+            sinsemilla_config.perm.clone(),
+        );
+
+        let l_star_plus1 = meta.fixed_column();
+
+        // Check that pieces have been decomposed correctly for Sinsemilla hash.
+        // <https://zips.z.cash/protocol/nu5.pdf#orchardmerklecrh>
+        //
+        // a = a_0||a_1 = l_star || (bits 0..=239 of left)
+        // b = b_0||b_1||b_2 = (bits 240..=244 of left) || (bits 245..=254 of left) || (bits 0..=234 of right)
+        // c = bits 235..=254 of right
+        //
+        // The message pieces `a`, `b`, `c` are constrained by Sinsemilla to be
+        // 250 bits, 250 bits, and 20 bits respectively.
+        //
+        meta.create_gate("Decomposition check", |meta| {
+            let a_whole = meta.query_advice(advices[0], Rotation::cur());
+            let b_whole = meta.query_advice(advices[1], Rotation::cur());
+            let c_whole = meta.query_advice(advices[2], Rotation::cur());
+            let left_node = meta.query_advice(advices[3], Rotation::cur());
+            let right_node = meta.query_advice(advices[4], Rotation::cur());
+
+            // z_1 of SinsemillaHash(a) = a_1
+            let z_a1 = meta.query_advice(advices[0], Rotation::next());
+            let a_1 = z_a1;
+            // a_0 = a - (a_1 * 2^10)
+            let a_0 = a_whole - a_1.clone() * pallas::Base::from_u64(1 << 10);
+
+            // b_0 has been constrained to be 5 bits outside this gate.
+            let b_0 = meta.query_advice(advices[1], Rotation::next());
+
+            // b_1 is expected to be a 10-bit value.
+            let b_1 = meta.query_advice(advices[2], Rotation::next());
+            let b1_z1 = meta.query_advice(advices[3], Rotation::next());
+
+            // b_2 = (b - (b_0 + 2^5 b_1)) / (2^15)
+            let b_2 = (b_whole - (b_0.clone() + b_1.clone() * pallas::Base::from_u64(1 << 5)))
+                * pallas::Base::TWO_INV.pow(&[15, 0, 0, 0]);
+
+            // Check that b_1 is a 10-bit value, i.e. b1_z1 = 0.
+            let b1_range_check = b1_z1;
```

Comment:
@daira and I discussed a general solution: range-constrain each subpiece individually. This would also involve redistributing pieces larger than 10 bits, to make full use of Sinsemilla's free range-constraint.